src/mlpack/tests/rbm_network_test.cpp

```diff
@@ -0,0 +1,250 @@
+/**
+ * @file rbm_network_test.cpp
+ * @author Kris Singh and Shikhar Jaiswal
+ *
+ * Tests the RBM Network
+ *
+ * digits dataset source:
+ * @misc{Lichman:2013 ,
+ * author = "M. Lichman",
+ * year = "2013",
+ * title = "{UCI} Machine Learning Repository",
+ * url = "http://archive.ics.uci.edu/ml",
+ * institution = "University of California,
+ * Irvine, School of Information and Computer Sciences" } 
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <mlpack/core.hpp>
+
+#include <mlpack/core/optimizers/rmsprop/rmsprop.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/rbm/binary_rbm_policy.hpp>
+#include <mlpack/methods/ann/rbm/spike_slab_rbm_policy.hpp>
+#include <mlpack/methods/ann/rbm.hpp>
+#include <mlpack/methods/softmax_regression/softmax_regression.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+#include <mlpack/core/optimizers/lbfgs/lbfgs.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::regression;
+
+BOOST_AUTO_TEST_SUITE(RbmNetworkTest);
+BOOST_AUTO_TEST_CASE(ClassificationTest)
+{
+  // Normalised dataset.
+  int hiddenLayerSize = 100;
+  size_t batchSize = 10;
+  size_t numEpoches = 30;
+  arma::mat trainData, testData, dataset;
+  arma::mat trainLabelsTemp, testLabelsTemp;
+  trainData.load("digits_train.arm");
+  testData.load("digits_test.arm");
+  trainLabelsTemp.load("digits_train_label.arm");
+  testLabelsTemp.load("digits_test_label.arm");
+
+  arma::Row<size_t> trainLabels = arma::zeros<arma::Row<size_t>>(1,
+      trainLabelsTemp.n_cols);
+  arma::Row<size_t> testLabels = arma::zeros<arma::Row<size_t>>(1,
+      testLabelsTemp.n_cols);
+
+  for (size_t i = 0; i < trainLabelsTemp.n_cols; ++i)
+    trainLabels(i) = arma::as_scalar(trainLabelsTemp.col(i));
+
+  for (size_t i = 0; i < testLabelsTemp.n_cols; ++i)
+    testLabels(i) = arma::as_scalar(testLabelsTemp.col(i));
+
+  arma::mat output, XRbm(hiddenLayerSize, trainData.n_cols),
+      YRbm(hiddenLayerSize, testLabels.n_cols);
+
+  XRbm.zeros();
+  YRbm.zeros();
+
+  BinaryRBMPolicy<> binary_rbm(trainData.n_rows, hiddenLayerSize);
+  GaussianInitialization gaussian(0, 0.1);
+  RBM<GaussianInitialization, BinaryRBMPolicy<>> model(trainData,
+      gaussian, binary_rbm, 1,  1, true, false);
+
+  size_t numRBMIterations = trainData.n_cols * numEpoches;
+  numRBMIterations /= batchSize;
+
+  StandardSGD msgd(0.06, batchSize, numRBMIterations, 0, true);
+  model.Reset();
+  model.Policy().VisibleBias().ones();
+  model.Policy().HiddenBias().ones();
+  // Test the reset function.
+  model.Train(trainData, msgd);
```

Comment:
@rcurtin Was the optimizer API changed recently? This seems to be failing to build here as follows:
```
/home/shikhar/mlpack/src/mlpack/core/optimizers/sgd/sgd_impl.hpp:118:22: error: no matching function for call to ‘mlpack::optimization::Function<mlpack::ann::RBM<mlpack::ann::GaussianInitialization, mlpack::ann::BinaryRBMPolicy<> > >::EvaluateWithGradient(arma::mat&, size_t&, arma::mat&, const size_t&)’
     overallObjective += f.EvaluateWithGradient(iterate, currentFunction,
```

Reply:
Yes, for more information see: https://github.com/mlpack/mlpack/pull/1252/ realistically you just have to add

```c++
double Evaluate(const arma::mat& parameters,
                const size_t begin,
                const size_t batchSize)
{
   return Evaluate(parameters, begin, batchSize, true);
}
```

see for an example: https://github.com/mlpack/mlpack/pull/1252/files#diff-a778189f4c83e20d3e29c6c03039abfd

Reply:
@zoq I've updated the changes here, but it still doesn't fix the issue.